## Task 1

In [23]:
import pandas as pd
import numpy as np
print(pd.__version__)

2.3.3


## Task2

In [ ]:
customers = pd.read_csv("data/customers.csv")
customers

## Task3

In [ ]:
customers.head(6)

In [ ]:
customers.tail(4)

In [ ]:
customers.sample(5)

## Task4

In [ ]:
cars = pd.read_csv("data/cars.csv")
cars.info()


- 398 Rows
- 9 Columns
- MPG, Cylinders, Displacement, Horsepower, Weight, Acceleration, Model Year, Origin, Car Name

## Task5
- Whole Numbers: Cylinders, Weight, Model Year, Origin
- Floats: MPG, Displacement, Acceleration
- Objects: Horsepower, Car Name

## Task6

In [38]:
titanic = pd.read_csv("data/titanic_survival.csv")
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


- 891 Rows, 15 columns
- Most missing values are in "deck"
- You know a value is missing by the "Non-Null Count" column, where the difference between the non-null count and the index range is the amount of missing values.

## Task7

In [21]:
cars.describe()

,mpg,cylinders,displacement,weight,acceleration,model year,origin
count,398.000000,398.000000,398.000000,398.000000,398.000000,398.000000,398.000000
mean,23.514573,5.454774,193.425879,2970.424623,15.568090,76.010050,1.572864
std,7.815984,1.701004,104.269838,846.841774,2.757689,3.697627,0.802055
min,9.000000,3.000000,68.000000,1613.000000,8.000000,70.000000,1.000000
25%,17.500000,4.000000,104.250000,2223.750000,13.825000,73.000000,1.000000
50%,23.000000,4.000000,148.500000,2803.500000,15.500000,76.000000,1.000000
75%,29.000000,8.000000,262.000000,3608.000000,17.175000,79.000000,2.000000
max,46.600000,8.000000,455.000000,5140.000000,24.800000,82.000000,3.000000


- 23.51
- Displacement: 455.0
- Median Weight: 2803

## Task 8

In [ ]:
print(titanic["fare"].mean())
print(titanic["fare"].median())
print(titanic["fare"].std())
print(titanic["fare"].min())
print(titanic["fare"].max())

## Task 9

In [32]:
data ={
    "Product": ["Shovel", "Rake", "Wheelbarrow", "Pressure Washer"],
    "Price": [30.00, 25.00, 50.00, 150.00],
    "In Stock": [True, False, True, True]
}
newdf = pd.DataFrame(data)
print(newdf)

           Product  Price  In Stock
0           Shovel   30.0      True
1             Rake   25.0     False
2      Wheelbarrow   50.0      True
3  Pressure Washer  150.0      True


## Task 10
- shape gives you an outline of the size of your dataframe, while info gives you a detailed breakdown of the types of information, missing values, etc.

- .head() gives you the first few entries, while .sample(). Sample can be useful if data is skewed at the start of your dataset, giving you a more accurate representation of the average values.

- It depends on what deck represents, if it's trying to say what deck individuals were on at the time of the incident, there wouldn't be a tangible record and you'd be relying on accounts from survivors who went through a traumatic event.

## Task 11

In [42]:
titanic["age"].describe()
titanic.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


- Youngest age was just under half a year old
- Oldest age was 80
- The count row will give you a count of the entries, it may not be the same number as the total number of rows, as it won't count None/NaN

## Task 12

In [44]:
cars.info()
print("")
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   age               300 non-null    float64
 1   income            300 non-null    float64
 2   visits_per_month  300 non-null    float64
dtypes: float64(3)
m

- Cars has more rows
- Cars has more columns
- They don't have missing values, however, Horsepower is of type object, which could reflect entries that don't align with expectations.
- Customers would likely be easier to start analyzing, as it is significantly smaller. However, with fewer columns, you likely would be able to find more practical information from Cars.

## Task 13

In [ ]:
cars.sample(5)

- Sample with a random_state will continually return the same indexes
- In case you would like compare multiple samples with each other, or return to the same random sample later for further evaluation.

## Task 14

In [ ]:
def load_and_check(filepath):
   new_file = pd.read_csv(filepath)
   print(f" Loaded: {new_file.shape[0]} rows, {new_file.shape[1]} columns.")
   return new_file

load_and_check("data/customers.csv")
load_and_check("data/titanic_survival.csv")
load_and_check("data/cars.csv")

## Task 15

In [62]:
cars["horsepower"].sample(20)

245     66
39     175
1      165
384     67
30      90
220     70
336      ?
285    130
186     83
92     158
164    110
47     100
119     91
98     100
182     86
376     68
208    150
253     95
100     88
49      86
Name: horsepower, dtype: object

- During one of the random samplings, one of the values was "?", likely used as a placeholder by whoever inputted the data for a value they didn't know.

## Task 16

In [66]:
df = pd.read_csv("data/titanic_survival.csv")
print(df.shape)
df.head(5)
df.info()
df.describe()

(891, 15)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


- We can see from the shape of the set, that this is a detailed set with 15 features and 891 observations. Based on the columns, and filename, it appears to be a dataset reflecting information about passengers on the Titanic, including identifiable information and information related to their deaths.